File: packages/core/src/types/consumer.ts
/**
 * @fileoverview Consumer types for anyq
 * @module @anyq/core/types/consumer
 */

import type { IMessage } from './message.js';
import type { HealthStatus } from './producer.js';

/**
 * Message handler function
 */
export type MessageHandler<T = unknown> = (
  message: IMessage<T>
) => Promise<void>;

/**
 * Batch message handler function
 */
export type BatchMessageHandler<T = unknown> = (
  messages: IMessage<T>[]
) => Promise<void>;

/**
 * Options for subscribing to messages
 */
export interface SubscribeOptions {
  /** Start consuming from the beginning (Kafka, NATS) */
  fromBeginning?: boolean;

  /** Start from specific timestamp (Kafka, Google Pub/Sub, NATS) */
  fromTimestamp?: Date;

  /** Maximum concurrent message processing. Default: 1 */
  concurrency?: number;

  /** Auto-acknowledge after handler completes without error. Default: true */
  autoAck?: boolean;

  /** Maximum messages per batch (for batch handlers). Default: 10 */
  batchSize?: number;

  /** Maximum time to wait for batch to fill in ms. Default: 1000 */
  batchTimeout?: number;
}

/**
 * Seek position for repositioning consumer
 */
export interface SeekPosition {
  /** Seek to beginning of queue/topic */
  beginning?: boolean;

  /** Seek to end (only new messages) */
  end?: boolean;

  /** Seek to specific timestamp */
  timestamp?: Date;

  /** Seek to specific offset (Kafka) */
  offset?: string;

  /** Partition to seek (Kafka) */
  partition?: number;
}

/**
 * Partition lag details
 */
export interface PartitionLag {
  partition: number;
  currentOffset: string;
  highWatermark: string;
  lag: number;
}

/**
 * Consumer lag metrics
 */
export interface ConsumerLag {
  /** Total lag across all partitions/streams */
  totalLag: number;

  /** Per-partition lag details (Kafka) */
  partitions?: PartitionLag[];
}

/**
 * Consumer events
 */
export interface ConsumerEvents<T = unknown> {
  /** Emitted on errors */
  error: (error: Error) => void;

  /** Emitted when backpressure is triggered */
  backpressure: (info: { active: number; highWaterMark: number }) => void;

  /** Emitted during rebalance (Kafka) */
  rebalancing: (info: { partitions: number[] }) => void;

  /** Emitted after rebalance completes (Kafka) */
  rebalanced: (info: { partitions: number[] }) => void;

  /** Emitted when consumer crashes */
  crash: (error: Error) => void;

  /** Emitted when message is received (before processing) */
  message: (message: IMessage<T>) => void;
}

/**
 * Event emitter interface for consumers
 */
export interface IConsumerEventEmitter<T = unknown> {
  on<K extends keyof ConsumerEvents<T>>(
    event: K,
    listener: ConsumerEvents<T>[K]
  ): this;

  off<K extends keyof ConsumerEvents<T>>(
    event: K,
    listener: ConsumerEvents<T>[K]
  ): this;

  emit<K extends keyof ConsumerEvents<T>>(
    event: K,
    ...args: Parameters<ConsumerEvents<T>[K]>
  ): boolean;
}

/**
 * Consumer interface for receiving messages
 *
 * @template T - Type of message body
 */
export interface IConsumer<T = unknown> extends IConsumerEventEmitter<T> {
  /**
   * Connect to the message broker
   *
   * @throws {ConnectionError} If connection fails after retries
   */
  connect(): Promise<void>;

  /**
   * Disconnect from the message broker
   */
  disconnect(): Promise<void>;

  /**
   * Check if currently connected
   */
  isConnected(): boolean;

  /**
   * Subscribe to messages with individual handler
   *
   * @param handler - Function to process each message
   * @param options - Subscription options
   */
  subscribe(
    handler: MessageHandler<T>,
    options?: SubscribeOptions
  ): Promise<void>;

  /**
   * Subscribe to messages with batch handler
   *
   * @param handler - Function to process message batches
   * @param options - Subscription options
   */
  subscribeBatch(
    handler: BatchMessageHandler<T>,
    options?: SubscribeOptions
  ): Promise<void>;

  /**
   * Pause message consumption
   */
  pause(): Promise<void>;

  /**
   * Resume message consumption
   */
  resume(): Promise<void>;

  /**
   * Check if consumption is paused
   */
  isPaused(): boolean;

  /**
   * Seek to specific position (Kafka, NATS)
   *
   * @param position - Position to seek to
   */
  seek?(position: SeekPosition): Promise<void>;

  /**
   * Get consumer lag metrics (Kafka)
   */
  getLag?(): Promise<ConsumerLag>;

  /**
   * Get consumer health status
   */
  healthCheck(): Promise<HealthStatus>;
}

File: go/core/consumer.go
package core

import (
	"context"
	"time"
)

// Handler processes a single message. Returning an error triggers the consumer's
// configured Strategy (or the adapter's legacy behaviour when no strategy is set).
type Handler func(ctx context.Context, msg Message) error

// BatchHandler processes a batch of messages.
type BatchHandler func(ctx context.Context, msgs []Message) error

// SubscribeOptions configure a subscription.
type SubscribeOptions struct {
	// FromBeginning starts consuming from the beginning (Kafka, NATS).
	FromBeginning bool
	// FromTimestamp starts from a specific time (Kafka, Pub/Sub, NATS).
	FromTimestamp *time.Time
	// Concurrency is the max concurrent message processing. Default: 1.
	Concurrency int
	// AutoAck auto-acknowledges after a handler returns nil. Default: true.
	AutoAck *bool
	// BatchSize is the max messages per batch (batch handlers). Default: 10.
	BatchSize int
	// BatchTimeout is the max wait to fill a batch. Default: 1s.
	BatchTimeout time.Duration
}

// AutoAckEnabled reports the resolved AutoAck value (defaulting to true).
func (o *SubscribeOptions) AutoAckEnabled() bool {
	if o == nil || o.AutoAck == nil {
		return true
	}
	return *o.AutoAck
}

// SeekPosition repositions a consumer (Kafka, NATS).
type SeekPosition struct {
	Beginning bool
	End       bool
	Timestamp *time.Time
	Offset    string
	Partition *int
}

// PartitionLag describes lag for one partition.
type PartitionLag struct {
	Partition     int
	CurrentOffset string
	HighWatermark string
	Lag           int64
}

// ConsumerLag describes total and per-partition lag.
type ConsumerLag struct {
	TotalLag   int64
	Partitions []PartitionLag
}

// Consumer receives messages from a broker.
type Consumer interface {
	Connect(ctx context.Context) error
	Disconnect(ctx context.Context) error
	IsConnected() bool

	// Subscribe consumes messages with a per-message handler. It blocks until
	// ctx is cancelled or a fatal error occurs.
	Subscribe(ctx context.Context, handler Handler, opts *SubscribeOptions) error
	// SubscribeBatch consumes messages with a batch handler.
	SubscribeBatch(ctx context.Context, handler BatchHandler, opts *SubscribeOptions) error

	Pause(ctx context.Context) error
	Resume(ctx context.Context) error
	IsPaused() bool

	HealthCheck(ctx context.Context) (HealthStatus, error)
}

// Seeker is implemented by consumers that support repositioning.
type Seeker interface {
	Seek(ctx context.Context, pos SeekPosition) error
}

// Lagger is implemented by consumers that report lag (Kafka).
type Lagger interface {
	GetLag(ctx context.Context) (ConsumerLag, error)
}
